ddtrace/contrib/flask/patch.py

```diff
@@ -1,24 +1,481 @@
+
+        # Wrap the `start_response` handler to extract response code
+        # DEV: We tried using `Flask.finalize_request`, which seemed to work, but gave us hell during tests
+        # DEV: The downside to using `start_response` is we do not have a `Flask.Response` object here,
+        #   only `status_code`, and `headers` to work with
+        #   On the bright side, this works in all versions of Flask (or any WSGI app actually)
+        def _wrap_start_response(func):
+            def traced_start_response(status_code, headers):
+                code, _, _ = status_code.partition(' ')
+                try:
+                    code = int(code)
+                except ValueError:
+                    pass
+
+                # Override root span resource name to be `<method> 404` for 404 requests
+                # DEV: We do this because we want to make it easier to see all 404 requests together
+                #      Also, we do this to reduce the cardinality on unknown urls
+                if code == 404:
+                    s.resource = u'{} 404'.format(request.method)
+
+                s.set_tag(http.STATUS_CODE, code)
+                if code in config.flask.get('error_codes', set()):
+                    s.error = 1
+                return func(status_code, headers)
+            return traced_start_response
+        start_response = _wrap_start_response(start_response)
+
+        # DEV: We set response status code in `_wrap_start_response`
+        s.set_tag(http.URL, request.url)
+        s.set_tag(http.METHOD, request.method)
+
+        # TODO: Add request header tracing
+        # for k, v in request.headers:
+        #     s.set_tag('http.request.headers.{}'.format(k), v)
+        return wrapped(environ, start_response)
+
+
+def traced_blueprint_register(wrapped, instance, args, kwargs):
+    """
+    Wrapper for flask.blueprints.Blueprint.register
+
+    This wrapper just ensures the blueprint has a pin, either set manually on
+    itself from the user or inherited from the application
+    """
+    def _wrap(app, *args, **kwargs):
+        pin = Pin.get_from(instance)
+        if not pin:
+            pin = Pin.get_from(app)
+            if pin:
+                pin.clone().onto(instance)
+        return wrapped(app, *args, **kwargs)
+    return _wrap(*args, **kwargs)
+
+
+def traced_blueprint_add_url_rule(wrapped, instance, args, kwargs):
+    pin = get_inherited_pin(wrapped, instance)
+    if not pin:
+        return wrapped(*args, **kwargs)
+
+    def _wrap(rule, endpoint=None, view_func=None, **kwargs):
+        if view_func:
+            pin.clone().onto(view_func)
+        return wrapped(rule, endpoint=endpoint, view_func=view_func, **kwargs)
+
+    return _wrap(*args, **kwargs)
+
+
+def traced_add_url_rule(wrapped, instance, args, kwargs):
+    """Wrapper for flask.app.Flask.add_url_rule to wrap all views attached to this app"""
+    def _wrap(rule, endpoint=None, view_func=None, **kwargs):
```

Comment:
@Kyle-Verhoog you left a comment about one of these wrapper functions before, but the comments got lost when I renamed the file.


I think I actually prefer this approach to the alternative and I find any of the overhead of creating/calling a function worth while compared to not doing it this way.

In a lot of these functions we are extracting a function from the argument list, wrapping, and then passing to the original function call.

Without this helper function we end up with:

```python
def traced_blueprint_add_url_rule(wrapped, instance, args, kwargs):
    pin = get_pin()
    
    rule = kwargs.get('rule')
    if not rule and len(args) > 0:
        rule = args[0]

    endpoint = kwargs.get('endpoint')
    if not endpoint and len(args) > 1:
        endpoint = args[1]

    view_func = kwargs.get('view_func')
    if not view_func and len(args) > 2:
        view_func = args[2]

    if view_func:
        view_func = wrap_function(instance, view_func, name=endpoint, resource=rule)
        # The value we extract from `args` or `kwargs` isn't always a reference,
        #   so we have to write it back in to where we grabbed it from. :/
        if len(args) > 2:
            args = args[0:2] + (view_func, ) args[3:]
        else:
            kwargs['view_func'] = view_func

    return wrapped(*args, **kwargs)
```

I am still open to the discussion on whether having a helper `def _wrap()` is good for us or not, but IMHO it makes the development here 100x easier.

Reply:
Oh I see! Okay, I get it for when there's an argument we want to intercept and replace. The original comment I left was for the `traced_blueprint_register` (and similar) functions which didn't seem to require any argument fiddling but still had the `_wrap()` approach. This seemed excessive to me.

It's not a show-stopper by any means, it just added some indirection which made it harder for me to follow when trying to figure out what was going on.

Reply:
I must have missed noticing that. I will go back through and make sure we are only using `_wrap()` when it is appropriate. :+1: